node/src/components/consensus/highway_core/state.rs

```diff
@@ -392,7 +426,23 @@ impl<C: Context> State<C> {
         // `cumulative_w[i]` denotes the last weight unit that belongs to validator `i`.
         // `binary_search` returns the first `i` with `cumulative_w[i] >= r`, i.e. the validator
         // who owns the randomly selected weight unit.
-        self.cumulative_w.binary_search(&r).unwrap_or_else(identity)
+        let leader_index = self.cumulative_w.binary_search(&r).unwrap_or_else(|| {
+            error!("random number out of range");
+            ValidatorIndex(0)
```

Comment:
Is that a safe thing to do? Wouldn't it, in the worst case, keep assigning the slot to the same validator? I would prefer to err on the safe side and exit the node.

Reply:
Unless there is a major bug in `ValidatorMap::binary_search` — which we'd probably see in both the unit and integration tests —, this can never happen, so I was even considering `unwrap` here.

But I thought I'd err on the safer side and only log an error: Even if we'd assign lots of slots to the same validator (and fixing it in the next upgrade), that's still better than shutting down the whole network.

Also, passing this unreachable special case up through the whole call hierarchy would complicate everything. These methods should be infallible.

Reply:
> Even if we'd assign lots of slots to the same validator (and fixing it in the next upgrade), that's still better than shutting down the whole network.

Unless this is a result of an attack where a malicious actor managed to force the node to go into the `unwrap_or_else` case while becoming the validators at index 0.

Reply:
But unless there is a huge bug in the very short `binary_search` implementation there's no way for an attacker to do this.

Reply:
I'm torn here :P On the one hand, if this should be unreachable, I'd prefer the code to reflect that, so just `unwrap()` here. On the other hand, if we're mistaken, this could bring the whole network down (at least temporarily).
I guess the question here is how likely is it that we would miss the error message for a longer period? If it's unlikely, then I'd lean slightly towards Andreas' solution (better to have the network running if the error doesn't cause huge issues, and this one shouldn't). If it's likely that the errors would go unnoticed for longer, I'd say better to force us to notice by crashing.
I think it's rather unlikely, though (we also have alerts, don't we?), so I think I'd go with Andreas' solution.

Reply:
If it's possible to recover and keep the network live, we should always do so.